### PR TITLE
WIP: tree vectorizing transformation

### LINKED ISSUE #1012: pytree transformation
I'd like to write code like `np.sum((x - y) ** 2)` that automatically works on arbitrarily nested Python objects.

I think this could be done cleanly with a "pytree transformation" that acts on functions, somewhat similar to the existing `vmap`, e.g.,
```
@jax.autotree
def norm(x, y):
    return np.sum((x - y) ** 2)
```

Under the hood, we would need to define what "autotree" versions of each primitive function look like on a list of heterogeneously shaped arrays, e.g.,
- Vectorized functions like `+` operate over each array independently.
- In mixed tree/array operations, the array should get "broadcast" by repeating it when the structure of the tree, e.g., `{'a': x, 'b': y'} + z` -> `{'a': x + z, 'b': y + z}`.
- Reductions like `sum` sum over all the arrays to produce a scalar, unless an axis is specified.

The upside is that this could make it trivial to write functional transformations like `jax.experimental.optimizers` in a fully general way. I can think of lots of nice applications for this, e.g., for methods that solve equations.

### PR DESCRIPTION
**Pull request by shoyer**
Fixes https://github.com/google/jax/issues/1012

The idea behind this transformation is that every argument to a function decorated by `@tree_vectorize` is virtually flattened and concatenated into 1D vector, e.g., `{'x': 1, 'y': array([[2, 3]])}` becomes `array([1, 2, 3])`.

However, under the covers the original arrays are still preserved with their original shapes. This has several advantages:

1. we don't actually need to copy the arrays into a giant vector (avoiding the unnecessary memory copies)
2. all operations happen on arrays with the original shapes (useful for predictable performance on TPUs)
3. we can restore the tree structure on the outputs of the function

The hope is that this should make it considerably easier to write efficient numerical algorithms that work on pytrees (e.g., for optimization, linear equation solving or ODE solving).

Here are a few simple examples:
```python
from jax import tree_vectorize
import jax.numpy as jnp

x = {'a': 1, 'b': jnp.array([2, 3])}
y = 1
add = tree_vectorize(jnp.add)
print(add(x, y))
# {'a': DeviceArray(2, dtype=int32), 'b': DeviceArray([3, 4], dtype=int32)}

@tree_vectorize
def norm(x):
  assert x.shape == (3,)
  return jnp.sqrt(jnp.sum(x ** 2))

print(norm(x))
# 3.7416575
```

Even though the inputs are always treated as vectors, outputs can have any number of dimensions, e.g., producing nested pytrees-of-pytrees, as shown in this example from the unit tests:
```python
@tree_vectorize
@shapecheck(['n', 'm'], '(n, m)')  # just to show that it works!
def add_outer(x, y):
  return jnp.expand_dims(x, 1) + jnp.expand_dims(y, 0)

tree1 = {'a': 1, 'b': jnp.array([2, 3])}
tree2 = {'c': 10, 'd': jnp.array([20, 30])}
pprint.pprint(add_outer(tree1, tree2))
# {'a': {'c': DeviceArray(11, dtype=int32),
#        'd': DeviceArray([21, 31], dtype=int32)},
#  'b': {'c': DeviceArray([12, 13], dtype=int32),
#        'd': DeviceArray([[22, 32],
#              [23, 33]], dtype=int32)}}
```

TODO:

- [ ] missing ops for `cg`:
    - [x] `jit`
    - [ ] `while_loop`
- [ ] further missing ops for `odeint`:
    - [x] `concatenate`
    - [ ] `gather`
    - [ ] `scatter`
- [ ] measure performance improvement for `odeint`!
- [ ] figure out the API, e.g.,
     - [ ] explicit API for indicating functions, especially with multiple outputs?
     - [ ] explicit API for indicating non-vectorized arguments?
- [ ] make things more robust
     - [ ] `tree_call` works with `vmap`, `jvp` and `vjp` (i.e., to implement `jacfwd` and `jacrev`)
     - [x] missing `Trace` methods like `process_call`, `post_process_call`
     - [x] support matching against explicit axes with trivial treedefs, e.g., so `x + jnp.zeros(...)` and explicit broadcasting work. 
- [ ] more comprehensive tests
     - [ ] check exceptions
     - [ ] make a reference implementation for a subset of `tree_vectorize` use cases that works by explicitly unraveling/concatenating

### CODE REVIEW

**Comment by shoyer** [comment]
A good test for composability are the various autodiff helper functions like `jacfwd` and `jacrev`.

In theory, the following should work, using `tree_vectorize` to handle all of the tree-mangling from `api.py` (`_std_basis`, `_unravel_array_into_pytree`, `tree_transpose`, etc):
```python
  def test_jacobian(self):

    @tree_vectorize
    def _jacfwd(f, x):
      pushfwd = partial(jax.jvp, f, (x,))
      basis = jnp.eye(x.size, dtype=x.dtype)
      y, jac = jax.vmap(pushfwd, out_axes=(None, 1))((basis,))
      return jac

    @tree_vectorize
    def _jacrev(f, x):
      y, pullback = jax.vjp(f, x)
      basis = jnp.eye(y.size, dtype=y.dtype)
      jac = jax.vmap(pullback)(basis)
      return jac

    def _apply_argnums(transform, fun, argnums=0):
      def jacfun(*args):
        f_partial, dyn_args = api.argnums_partial(
            lu.wrap_init(fun), argnums, args)
        result = transform(f_partial.call_wrapped, dyn_args)
        return result[0] if isinstance(argnums, int) else result
      return jacfun

    jacfwd = partial(_apply_argnums, _jacfwd)
    jacrev = partial(_apply_argnums, _jacrev)

    f = lambda x: {'c': x['a'] * (1 + x['b'] ** 2), 'd': x['a'] - x['b']}
    tree = {'a': 1.0, 'b': 2.0}
    expected = jax.jacfwd(f)(tree)

    actual = jacrev(f)(tree)
    self.assertTreeEqual(expected, actual)

    actual = jacrev(f)(tree)
    self.assertTreeEqual(expected, actual)
```

However, this test currently fails, because the way I handle function arguments with `tree_call` doesn't seem to work properly with autodiff. I think the problem is that `tree_call` needs to work more like an actual `call` primitive (re-invoking tracing), but there may be other ways to fix it.

Attention from somebody who deeply understands JAX's transform rules (i.e., @mattjj ) would be appreciated :)

**Comment by shoyer** [comment]
I simplified the issue from `jacobian` into `test_vmap_tree_call`. The underlying issue is that `tree_call` should commute with operations like batching, but that doesn't currently work.

File: tests/treevec_test.py [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+    return jacfun
+
+  @pytest.mark.xfail
+  def test_jacfwd(self):
+
+    @tree_vectorize
+    def _jacfwd(f, x):
+      pushfwd = partial(jax.jvp, f, (x,))
+      basis = jnp.eye(x.size, dtype=x.dtype)
+      y, jac = jax.vmap(pushfwd, out_axes=(None, 1))((basis,))
```
Comment (shoyer) [inline tests/treevec_test.py:501]: I think this is very close to working, except we don't handle `basis` properly yet because it has a different tree structure. (This works in other cases, but not for transformed tree_call functions yet.)

File: tests/treevec_test.py [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+    self.assertTreeEqual(actual, expected, check_dtypes=True)
+
+  @pytest.mark.xfail
+  def test_grad(self):
+    def g(x):
+      return x['a'] + 2 * x['b'] + x['c'] ** 2
+
+    @tree_vectorize
+    def f(g, x):
+      return jax.grad(g)(x)
```
Comment (shoyer) [inline tests/treevec_test.py:480]: Gradient support needs `JaxprTrace.process_tree_call` (which seems straightforward, but tedious).

File: jax/experimental/treevec.py [hunk 3]
```diff
@@ -0,0 +1,987 @@
...
+  primal_out, tangent_out = split_list(result, [num_out_primals()])
+  return [ad.JVPTracer(self, p, t) for p, t in zip(primal_out, tangent_out)]
+ad.JVPTrace.process_tree_call = _jvp_process_tree_call
+
+core.EvalTrace.process_tree_call = core.EvalTrace.process_call
+
+partial_eval.JaxprTrace.process_tree_call = partial_eval.JaxprTrace.process_call
+
+@lu.transformation_with_aux
+def _tree_batch_subtrace(main, in_dims_trees, *in_vals_trees, **params):
```
Comment (shoyer) [inline jax/experimental/treevec.py:333]: @apaszke  This is starting to feel pretty repetitive, especially when we consider doing the same thing for every higher order fucntions/transformation in JAX (e.g., partial eval, scan, while_loop, etc). It feels like there must be a better way to leverage existing support for wrapping trees...

**Comment by shoyer** [comment]
Closing this in favor of #8504
